File: front/types/ai.ts
/**
 * AI 模块类型定义
 */

// ============ AI 连接类型 ============

/** AI 服务提供商类型 */
export type AiConnectionType = 'ollama' | 'tensorrt' | 'openai' | 'claude' | 'qwen' | 'custom'

/** 代理类型 */
export type AiProxyType = 'http' | 'https' | 'socks5'

/** AI 代理配置 */
export interface AiProxyConfig {
  proxyType: AiProxyType
  host: string
  port: number
  username?: string
  password?: string
}

/** AI 连接配置 */
export interface AiConnection {
  id: string
  name: string
  connectionType: AiConnectionType
  host?: string
  port?: number
  apiKey?: string
  baseUrl?: string
  proxy?: AiProxyConfig
  sshConnectionId?: string
}

// ============ Ollama 类型 ============

/** Ollama 服务状态 */
export interface OllamaStatus {
  running: boolean
  version?: string
  host: string
  port: number
}

/** Ollama 模型详情 */
export interface OllamaModelDetails {
  format?: string
  family?: string
  parameterSize?: string
  quantizationLevel?: string
}

/** Ollama 模型信息 */
export interface OllamaModel {
  name: string
  size: number
  digest: string
  modifiedAt: string
  details?: OllamaModelDetails
}

/** 模型状态 */
export type ModelStatus = 'idle' | 'running' | 'downloading' | 'error'

/** Ollama 连接请求 */
export interface OllamaConnectRequest {
  host: string
  port: number
  sshConnectionId?: string
}

/** Ollama 运行中的模型 */
export interface OllamaRunningModel {
  name: string
  model: string
  size: number
  digest: string
  expiresAt?: string
  sizeVram?: number
}

/** 拉取模型进度 */
export interface OllamaPullProgress {
  status: string
  digest?: string
  total?: number
  completed?: number
}

// ============ GPU 监控类型 ============

/** GPU 信息 */
export interface GpuInfo {
  index: number
  name: string
  uuid: string
  driverVersion: string
  cudaVersion?: string
  architecture?: string
  computeCapability?: string
  memoryTotal: number    // MB
  memoryUsed: number     // MB
  memoryFree: number     // MB
  utilization: number    // %
  memoryUtilization: number // %
  temperature: number    // °C
  powerDraw?: number     // W
  powerLimit?: number    // W
  fanSpeed?: number      // %
}

/** GPU 进程信息 */
export interface GpuProcess {
  gpuUuid: string
  pid: number
  processName: string
  usedMemory: number     // MB
}

/** GPU 历史记录 */
export interface GpuHistory {
  timestamp: number
  gpuIndex: number
  utilization: number
  memoryUsed: number
  temperature: number
}

/** GPU 监控配置 */
export interface GpuMonitorConfig {
  enabled: boolean
  intervalSeconds: number
  retentionDays: number
  sshConnectionId?: string
}

/** 历史记录时间间隔 */
export type HistoryInterval = 'minute' | 'fiveminutes' | 'hour' | 'day'

// ============ AI 引擎标签类型 ============

/** AI 引擎类型 */
export type AiEngineType = 'ollama' | 'tensorrt' | 'cloudApi' | 'gpu' | 'mcp'

/** AI 模块状态 */
export interface AiModuleState {
  activeEngine: AiEngineType
  ollamaConnected: boolean
  ollamaConnectionId?: string
}

// ============ 云端 API 类型 ============

/** 云端 API 提供商 */
export type CloudApiProvider = 'openai' | 'claude' | 'qwen' | 'custom'

/** 云端 API 配置 */
export interface CloudApiConfig {
  id: string
  name: string
  provider: CloudApiProvider
  apiKey?: string
  baseUrl?: string
  organization?: string      // OpenAI organization
  defaultModel?: string
  proxy?: AiProxyConfig
  enabled: boolean
  createdAt?: string
  updatedAt?: string
}

/** 云端 API 模型信息 */
export interface CloudApiModel {
  id: string
  name: string
  provider: CloudApiProvider
  description?: string
  contextLength?: number
  pricing?: string
}

/** 云端 API 连接测试结果 */
export interface CloudApiTestResult {
  success: boolean
  message: string
  latencyMs?: number
}

/** 云端 API 保存请求 */
export interface CloudApiSaveRequest {
  id?: string               // undefined 表示创建，有值表示更新
  name: string
  provider: CloudApiProvider
  apiKey?: string
  baseUrl?: string
  organization?: string
  defaultModel?: string
  proxy?: AiProxyConfig
  enabled: boolean
}

/** 云端 API 测试请求 */
export interface CloudApiTestRequest {
  provider: CloudApiProvider
  apiKey: string
  baseUrl?: string
  organization?: string
  proxy?: AiProxyConfig
}

// ============ TensorRT LLM 类型 ============

/** TensorRT 模型信息 */
export interface TensorRTModel {
  id: string
  name: string
  version: string
  enginePath: string
  port: number
  status: ModelStatus
  gpuMemoryUsage?: number
  createdAt: string
}

/** TensorRT 模型部署配置 */
export interface TensorRTModelConfig {
  name: string
  modelPath: string           // HuggingFace 模型路径或本地路径
  maxBatchSize: number
  maxInputLen: number
  maxOutputLen: number
  quantization?: string       // fp16, int8, int4
  tensorParallelSize: number  // 多 GPU 并行
}

/** TensorRT 服务状态 */
export interface TensorRTStatus {
  installed: boolean
  version?: string
  cudaVersion?: string
  runningModels: number
}

/** TensorRT 连接请求 */
export interface TensorRTConnectRequest {
  host: string
  port?: number
  sshConnectionId?: string
}

// ============ MCP (Model Context Protocol) 类型 ============

/** MCP 传输类型 */
export type McpTransportType = 'stdio' | 'sse'

/** MCP Stdio 传输配置 */
export interface McpStdioTransport {
  type: 'stdio'
  command: string
  args: string[]
}

/** MCP SSE 传输配置 */
export interface McpSseTransport {
  type: 'sse'
  url: string
}

/** MCP 传输配置 */
export type McpTransport = McpStdioTransport | McpSseTransport

/** MCP Server 状态 */
export type McpServerStatus = 'running' | 'stopped' | { error: string }

/** MCP 工具定义 */
export interface McpTool {
  name: string
  description: string
  inputSchema: Record<string, unknown>
}

/** MCP Server 配置 */
export interface McpServerConfig {
  name: string
  description?: string
  transport: McpTransport
  tools: McpTool[]
}

/** MCP Server 信息 */
export interface McpServerInfo {
  id: string
  name: string
  description?: string
  transport: McpTransport
  status: McpServerStatus
  toolCount: number
  createdAt: string
}

/** MCP Server 创建请求 */
export interface McpServerCreateRequest {
  name: string
  description?: string
  transport: McpTransport
  tools?: McpTool[]
}

/** MCP 工具绑定请求 */
export interface McpToolBindRequest {
  serverId: string
  tool: McpTool
}

/** MCP 工具解绑请求 */
export interface McpToolUnbindRequest {
  serverId: string
  toolName: string
}

// ============ OpenWebUI 类型 ============

/** OpenWebUI 来源类型 */
export type OpenWebUISource = 'local' | 'docker' | 'notfound'

/** OpenWebUI 状态信息 */
export interface OpenWebUIStatus {
  available: boolean
  url?: string
  version?: string
  source: OpenWebUISource
}

// ============ 远程 AI 管理类型 ============

/** 远程 AI 环境信息 */
export interface RemoteAiEnvironment {
  ollamaInstalled: boolean
  ollamaVersion?: string
  tensorrtInstalled: boolean
  nvidiaGpuDetected: boolean
  gpuCount: number
}

// ============ 工具函数 ============

/** 格式化文件大小 */
export function formatModelSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`
}

/** 格式化模型修改时间 */
export function formatModelTime(isoTime: string): string {
  const date = new Date(isoTime)
  return date.toLocaleString()
}

File: backend/src/models/ai.rs
//! AI module data models
//!
//! This module contains data structures for AI service management,
//! including Ollama, TensorRT LLM, and cloud APIs.

use serde::{Deserialize, Serialize};

// ============ AI Connection Types ============

/// Proxy configuration for AI services
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProxyConfig {
    pub proxy_type: AiProxyType,
    pub host: String,
    pub port: u16,
    pub username: Option<String>,
    pub password: Option<String>,
}

/// Proxy type for AI services
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum AiProxyType {
    Http,
    Https,
    Socks5,
}

// ============ Ollama Models ============

/// Ollama service status
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OllamaStatus {
    pub running: bool,
    pub version: Option<String>,
    pub host: String,
    pub port: u16,
}

/// Ollama model information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OllamaModel {
    pub name: String,
    pub size: u64,
    pub digest: String,
    pub modified_at: String,
    #[serde(default)]
    pub details: Option<OllamaModelDetails>,
}

/// Ollama model details
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OllamaModelDetails {
    pub format: Option<String>,
    pub family: Option<String>,
    pub parameter_size: Option<String>,
    pub quantization_level: Option<String>,
}

/// Model status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ModelStatus {
    Idle,
    Running,
    Downloading,
    Error,
}

/// Ollama connect request
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OllamaConnectRequest {
    pub host: String,
    pub port: u16,
    pub ssh_connection_id: Option<String>,
}

// ============ Ollama API Response Types ============

/// Ollama API version response
#[derive(Debug, Clone, Deserialize)]
pub struct OllamaVersionResponse {
    pub version: String,
}

/// Ollama API tags (models) response
#[derive(Debug, Clone, Deserialize)]
pub struct OllamaTagsResponse {
    pub models: Vec<OllamaApiModel>,
}

/// Ollama API model info
#[derive(Debug, Clone, Deserialize)]
pub struct OllamaApiModel {
    pub name: String,
    pub size: u64,
    pub digest: String,
    pub modified_at: String,
    pub details: Option<OllamaApiModelDetails>,
}

/// Ollama API model details
#[derive(Debug, Clone, Deserialize)]
pub struct OllamaApiModelDetails {
    pub format: Option<String>,
    pub family: Option<String>,
    pub parameter_size: Option<String>,
    pub quantization_level: Option<String>,
}

impl From<OllamaApiModel> for OllamaModel {
    fn from(api_model: OllamaApiModel) -> Self {
        Self {
            name: api_model.name,
            size: api_model.size,
            digest: api_model.digest,
            modified_at: api_model.modified_at,
            details: api_model.details.map(|d| OllamaModelDetails {
                format: d.format,
                family: d.family,
                parameter_size: d.parameter_size,
                quantization_level: d.quantization_level,
            }),
        }
    }
}

/// Ollama running models response
#[derive(Debug, Clone, Deserialize)]
pub struct OllamaRunningModelsResponse {
    pub models: Option<Vec<OllamaRunningModel>>,
}

/// Ollama running model info
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OllamaRunningModel {
    pub name: String,
    pub model: String,
    pub size: u64,
    pub digest: String,
    // This struct is serialized to the frontend as camelCase but ALSO
    // deserialized from Ollama's /api/ps, which returns snake_case. The aliases
    // let deserialization accept the snake_case keys (otherwise these Options
    // silently stayed None) while serialization still emits camelCase.
    #[serde(alias = "expires_at")]
    pub expires_at: Option<String>,
    #[serde(alias = "size_vram")]
    pub size_vram: Option<u64>,
}

// ============ GPU Monitoring Models ============

/// GPU information from nvidia-smi
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GpuInfo {
    pub index: i32,
    pub name: String,
    pub uuid: String,
    pub driver_version: String,
    pub cuda_version: Option<String>,
    pub architecture: Option<String>,
    pub compute_capability: Option<String>,
    pub memory_total: u64,       // MB
    pub memory_used: u64,        // MB
    pub memory_free: u64,        // MB
    pub utilization: f32,        // %
    pub memory_utilization: f32, // %
    pub temperature: f32,        // °C
    pub power_draw: Option<f32>, // W
    pub power_limit: Option<f32>, // W
    pub fan_speed: Option<f32>,  // %
}

/// GPU process information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GpuProcess {
    pub gpu_uuid: String,
    pub pid: u32,
    pub process_name: String,
    pub used_memory: u64, // MB
}

/// GPU history record
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GpuHistory {
    pub timestamp: i64,
    pub gpu_index: i32,
    pub utilization: f32,
    pub memory_used: u64,
    pub temperature: f32,
}

// ============ Cloud API Models ============

/// Cloud API provider types
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum CloudApiProvider {
    OpenAI,
    Claude,
    Qwen,
    Custom,
}

impl std::fmt::Display for CloudApiProvider {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CloudApiProvider::OpenAI => write!(f, "OpenAI"),
            CloudApiProvider::Claude => write!(f, "Claude"),
            CloudApiProvider::Qwen => write!(f, "Qwen"),
            CloudApiProvider::Custom => write!(f, "Custom"),
        }
    }
}

/// Cloud API configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloudApiConfig {
    pub id: String,
    pub name: String,
    pub provider: CloudApiProvider,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub organization: Option<String>,      // OpenAI organization
    pub default_model: Option<String>,
    pub proxy: Option<AiProxyConfig>,
    pub enabled: bool,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

/// Cloud API model information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloudApiModel {
    pub id: String,
    pub name: String,
    pub provider: CloudApiProvider,
    pub description: Option<String>,
    pub context_length: Option<u64>,
    pub pricing: Option<String>,
}

/// Cloud API test connection result
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloudApiTestResult {
    pub success: bool,
    pub message: String,
    pub latency_ms: Option<u64>,
}

// ============ TensorRT LLM Models ============

/// TensorRT-LLM model information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TensorRTModel {
    pub id: String,
    pub name: String,
    pub version: String,
    pub engine_path: String,
    pub port: u16,
    pub status: ModelStatus,
    pub gpu_memory_usage: Option<u64>,
    pub created_at: String,
}

/// TensorRT-LLM model deployment configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TensorRTModelConfig {
    pub name: String,
    pub model_path: String,           // HuggingFace model path or local path
    pub max_batch_size: u32,
    pub max_input_len: u32,
    pub max_output_len: u32,
    pub quantization: Option<String>, // fp16, int8, int4
    pub tensor_parallel_size: u32,    // Multi-GPU parallelism
}

/// TensorRT-LLM service status
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TensorRTStatus {
    pub installed: bool,
    pub version: Option<String>,
    pub cuda_version: Option<String>,
    pub running_models: usize,
}

/// TensorRT connect request
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TensorRTConnectRequest {
    pub host: String,
    pub port: Option<u16>,
    pub ssh_connection_id: Option<String>,
}

// ============ MCP (Model Context Protocol) Models ============

/// MCP transport type
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum McpTransport {
    /// Stdio transport - runs command as subprocess
    Stdio {
        command: String,
        #[serde(default)]
        args: Vec<String>,
    },
    /// SSE transport - connects to HTTP endpoint
    Sse {
        url: String,
    },
}

/// MCP server configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServerConfig {
    pub name: String,
    pub description: Option<String>,
    pub transport: McpTransport,
    #[serde(default)]
    pub tools: Vec<McpTool>,
}

/// MCP server status
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum McpServerStatus {
    Running,
    Stopped,
    Error(String),
}

/// MCP server information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpServerInfo {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub transport: McpTransport,
    pub status: McpServerStatus,
    pub tool_count: usize,
    pub created_at: String,
}

/// MCP tool definition
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpTool {
    pub name: String,
    pub description: String,
    pub input_schema: serde_json::Value,
}

// ============ OpenWebUI Models ============

/// OpenWebUI source type
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum OpenWebUISource {
    /// Local installation (e.g., pip install)
    Local,
    /// Docker container
    Docker,
    /// Not found
    NotFound,
}

/// OpenWebUI status information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenWebUIStatus {
    pub available: bool,
    pub url: Option<String>,
    pub version: Option<String>,
    pub source: OpenWebUISource,
}

// ============ Remote AI Management Models ============

/// Remote AI environment information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteAiEnvironment {
    pub ollama_installed: bool,
    pub ollama_version: Option<String>,
    pub tensorrt_installed: bool,
    pub nvidia_gpu_detected: bool,
    pub gpu_count: usize,
}
